solutions/platinum/poi-10-frog.mdx

```diff
@@ -2,96 +2,171 @@
 id: poi-10-frog
 source: POI 2010
 title: Frog
-author: Benjamin Qi
+author: Benjamin Qi, Justin Ji
 ---
 
+## Explanation
+
+To calculate the $k$-th closest rock from a certain location,
+we maintain a sliding window of size $k+1$. Essentially, we sweep
+left to right, and keep on moving the window right until either:
+
+1. The window cannot be moved right any further.
+2. Moving the window to the right would increase the maximum distance from this current location.
+
+Afterwards, we can proceed to binary lift to get our answer. However, directly calculating binary lifts
+the traditional way would take up too much memory. From here, there are two ways to proceed:
+
+## Implementation 1
+
+Iterate over the power of two we are lifting on in increasing order. Similar to binary exponentiation,
+we calculate the $2^{i}$-th next location at every step, and check if we need to use this current lift.
+
+**Time Complexity:** $\mathcal{O}(N\log M)$
+
+**Space Complexity:** $\mathcal{O}(N)$
+
+<LanguageSection>
+<CPPSection>
 ```cpp
 #include <bits/stdc++.h>
-#include <ext/pb_ds/assoc_container.hpp>
-
 using namespace std;
-using namespace __gnu_pbds;
-
-typedef long long ll;
-typedef vector<int> vi;
-typedef pair<int, int> pii;
-template <class T>
-using Tree =
-    tree<T, null_type, less<T>, rb_tree_tag, tree_order_statistics_node_update>;
-
-#define FOR(i, a, b) for (int i = a; i < (b); i++)
-#define F0R(i, a) for (int i = 0; i < (a); i++)
-#define FORd(i, a, b) for (int i = (b) - 1; i >= a; i--)
-#define F0Rd(i, a) for (int i = (a) - 1; i >= 0; i--)
-
-#define sz(x) (int)(x).size()
-#define mp make_pair
-#define pb push_back
-#define f first
-#define s second
-#define lb lower_bound
-#define ub upper_bound
-#define all(x) x.begin(), x.end()
-
-const int MOD = 1000000007;
-const int MX = 1000001;
-
-int n, k, sz[MX], nex[MX][20];
-ll m;
-
-int get(int x, int lef) {
-	F0Rd(i, 20) if (lef & (1 << i)) x = nex[x][i];
-	return x;
-}
+using ll = long long;
 
-void solve(int x) {
-	if (m <= n) cout << get(x, m) << " ";
-	else {
-		x = get(x, n);
-		cout << get(x, (m - n) % sz[x]) << " ";
-	}
-}
+int main() {
+	cin.tie(0)->sync_with_stdio(0);
+	int n, k;
+	ll m;
+	cin >> n >> k >> m;
+	vector<ll> rocks(n);
+	for (ll &i : rocks) { cin >> i; }
 
-ll p[MX];
+	vector<int> next_stone(n);
+	// Maintain a sliding window of size k + 1 to
+	// calculate the next stone the frog will jump to.
+	int l = 0;
+	int r = k;
+	for (int i = 0; i < n; i++) {
+		while (r + 1 < n && rocks[r + 1] - rocks[i] < rocks[i] - rocks[l]) {
+			l++, r++;
+		}
+		next_stone[i] = (rocks[i] - rocks[l] >= rocks[r] - rocks[i]) ? l : r;
+	}
 
-void genNex() {
-	pii cur = {1, k + 1};
-	FOR(i, 1, n + 1) {
-		while (cur.s < n && p[cur.s + 1] - p[i] < p[i] - p[cur.f])
-			cur.f++, cur.s++;
-		if (p[i] - p[cur.f] >= p[cur.s] - p[i]) nex[i][0] = cur.f;
-		else nex[i][0] = cur.s;
+	const int LOG = (int)log2l(m) + 1;
```

Comment:
i mean u only use this once, why  not put it directly in the for loop with a <=

Reply:
true, lemme do that in a sec